aws/hll/aws-s3-transfer-manager/src/download.rs

```diff
@@ -34,17 +49,178 @@ impl From<GetObjectInputBuilder> for DownloadRequest {
+    ///         .bucket("my-bucket")
+    ///         .key("my-key")
+    ///         .into();
+    ///
+    ///     let handle = client.download(request).await?;
+    ///     // process data off handle...
+    ///     Ok(())
+    /// }
+    /// ```
+    pub async fn download(&self, req: DownloadRequest) -> Result<DownloadHandle, TransferError> {
+        // if there is a part number then just send the default request
+        if req.input.get_part_number().is_some() {
+            todo!("single part download not implemented")
+        }
+
+        let ctx = DownloadContext {
+            client: self.client.clone(),
+            target_part_size: self.target_part_size_bytes,
+        };
+
+        // make initial discovery about the object size, metadata, possibly first chunk
+        let mut discovery = discover_obj(&ctx, &req).await?;
+        let (comp_tx, comp_rx) = mpsc::channel(self.concurrency);
+        let start_seq = handle_discovery_chunk(&mut discovery, &comp_tx).await;
+
+        // spawn all work into the same JoinSet such that when the set is dropped all tasks are cancelled.
+        let mut tasks = JoinSet::new();
+
+        if !discovery.remaining.is_empty() {
+            // start assigning work
+            let (work_tx, work_rx) = async_channel::bounded(self.concurrency);
+            let input = req.input.clone();
+            let part_size = self.target_part_size_bytes;
+            let rem = discovery.remaining.clone();
+
+            tasks.spawn(distribute_work(rem, input, part_size, start_seq, work_tx));
+
+            for i in 0..self.concurrency {
+                let worker = chunk_downloader(ctx.clone(), work_rx.clone(), comp_tx.clone())
+                    .instrument(tracing::debug_span!("chunk-downloader", worker = i));
+                tasks.spawn(worker);
+            }
+        }
+
+        // drop our reference to completion sender such that when all workers drop theirs it's closed
+        drop(comp_tx);
+
+        let handle = DownloadHandle {
+            object_meta: discovery.meta,
+            body: Body::new(comp_rx),
+            tasks,
+        };
+
+        Ok(handle)
+    }
 }
 
-impl DownloadResponse {
-    /// Object metadata
-    pub fn object_meta(&self) -> &ObjectMetadata {
-        &self.object_meta
+/// Handle possibly sending the first chunk of data received through discovery. Returns
+/// the starting sequence number to use for remaining chunks.
+async fn handle_discovery_chunk(
+    discovery: &mut ObjectDiscovery,
+    completed: &mpsc::Sender<Result<ChunkResponse, TransferError>>,
+) -> u64 {
+    let mut start_seq = 0;
+    if let Some(initial_data) = mem::replace(&mut discovery.initial_chunk, None) {
+        let chunk = ChunkResponse {
+            seq: start_seq,
+            data: Some(initial_data),
+            object_meta: Some(discovery.meta.clone()),
+        };
+        completed.send(Ok(chunk)).await.expect("initial chunk");
+        start_seq += 1;
```

Comment:
No iteration happening here, prefer assignment:
```suggestion
        start_seq = 1;
```